src/Microsoft.DotNet.Interactive/KernelExtensions.cs

```diff
@@ -209,64 +211,90 @@ public static T UseDotNetVariableSharing<T>(this T kernel)
+        private static Command whos()
+        {
+            var command = new Command("#!whos", "Display the names of the current top-level variables and their values.")
+            {
+                Handler = CommandHandler.Create(async (ParseResult parseResult, KernelInvocationContext context) =>
+                {
+                    await  DisplayValues(context, true);
                 })
             };
 
-            // TODO: (who_and_whos) this should be a separate command with separate help
-            command.AddAlias("#!who");
-
             return command;
+        }
 
-            void Display(KernelInvocationContext context, bool detailed)
+        private static async Task DisplayValues(KernelInvocationContext context, bool detailed)
+        {
+            if (context.Command is SubmitCode &&
+                context.HandlingKernel is ISupportGetValue)
             {
-                if (context.Command is SubmitCode &&
-                    context.HandlingKernel is DotNetKernel kernel)
+                var nameEvents = new List<ValueInfosProduced>();
+
+                var result = await context.HandlingKernel.SendAsync(new RequestValueInfos(context.Command.TargetKernelName));
+                result.KernelEvents.OfType<ValueInfosProduced>().Subscribe(e => nameEvents.Add(e));
+
+                var valueNames = nameEvents.SelectMany(e => e.ValueInfos.Select(d => d.Name)).Distinct().ToList();
+
+                var valueEvents = new List<ValueProduced>();
+                var valueCommands = valueNames.Select(valueName => new RequestValue(valueName, context.HandlingKernel.Name)).ToList();
+
+
+
+                foreach (var valueCommand in valueCommands)
                 {
-                    var variables = kernel.GetVariableNames()
-                                          .Select(name =>
-                                          {
-                                              kernel.TryGetVariable(name, out object v);
-                                              return new CurrentVariable(name, v.GetType(), v);
-                                          });
-
-                    var currentVariables = new CurrentVariables(
-                        variables,
-                        detailed);
-
-                    var html = currentVariables
-                        .ToDisplayString(HtmlFormatter.MimeType);
-
-                    context.Publish(
-                        new DisplayedValueProduced(
-                            html,
-                            context.Command,
-                            new[]
-                            {
-                                new FormattedValue(
-                                    HtmlFormatter.MimeType,
-                                    html)
-                            }));
+                    result = await context.HandlingKernel.SendAsync(valueCommand);
+                    result.KernelEvents.OfType<ValueProduced>().Subscribe(e => valueEvents.Add(e));
```

Comment:
This should be disposed if we're being nitpicky about it.

Reply:
done